packages/modelence/src/lock/helpers.ts

```diff
@@ -0,0 +1,180 @@
+import { randomBytes } from 'crypto';
+import { locksCollection } from './collection';
+
+/**
+ * Unique identifier for this application instance.
+ * Generated once per application instance to track which container owns which locks.
+ */
+const containerId = randomBytes(32).toString('base64url');
+
+/**
+ * Time after which a lock is considered stale if no heartbeat update was received
+ */
+const LOCK_STALE_THRESHOLD_MS = 30000; // 30 seconds
+
+/**
+ * Interval at which lock heartbeats are updated in the database
+ */
+const LOCK_HEARTBEAT_INTERVAL_MS = 5000; // 5 seconds
+
+/**
+ * Map of lock types to their heartbeat interval timers
+ */
+const heartbeatIntervals = new Map<string, NodeJS.Timeout>();
+
+/**
+ * Acquires a lock of the specified type.
+ * If the lock already exists and is owned by another container, this will fail.
+ * If the lock is stale (no recent heartbeat), it will be taken over.
+ *
+ * @param type - The type of lock to acquire
+ * @param staleThreshold - Time in ms after which a lock is considered stale (default: 30s)
+ * @returns true if lock was acquired, false otherwise
+ */
+export async function acquireLock(
+  type: string,
+  staleThreshold: number = LOCK_STALE_THRESHOLD_MS
+): Promise<boolean> {
+  const now = new Date();
+  const staleThresholdDate = new Date(now.getTime() - staleThreshold);
+
+  console.log(`Attempting to acquire lock: ${type}`);
+
+  // Try to insert a new lock
+  try {
+    await locksCollection.insertOne({
+      type,
+      containerId,
+      acquiredAt: now,
+      heartbeatAt: now,
+    });
+    return true;
+  } catch {
+    // Lock already exists, check if it's stale
+    const existingLock = await locksCollection.findOne({
+      type,
+      heartbeatAt: { $lt: staleThresholdDate },
+    });
+
+    if (!existingLock) {
+      // Lock exists and is not stale
+      return false;
+    }
+
+    // Try to acquire the stale lock
+    const result = await locksCollection.updateOne(
+      {
+        type,
+        heartbeatAt: { $lt: staleThresholdDate },
+      },
+      {
+        $set: {
+          containerId,
+          acquiredAt: now,
+          heartbeatAt: now,
+        },
+      }
+    );
+
+    return result.modifiedCount > 0;
+  }
+}
+
+/**
+ * Releases a lock of the specified type owned by this container.
+ *
+ * @param type - The type of lock to release
+ * @returns true if lock was released, false if lock wasn't owned by this container
+ */
+export async function releaseLock(type: string): Promise<boolean> {
+  const result = await locksCollection.deleteOne({
+    type,
+    containerId,
+  });
+
+  return result.deletedCount > 0;
+}
+
+/**
+ * Updates the heartbeat timestamp for a lock owned by this container.
+ *
+ * @param type - The type of lock to update
+ * @returns true if heartbeat was updated, false if lock wasn't owned by this container
+ */
+async function updateLockHeartbeat(type: string): Promise<boolean> {
+  const result = await locksCollection.updateOne(
```

Comment:
Optional: can we keep a memory state whether we own the lock, and only attempt to update the heartbeat if we do? When an application scales to 10 or more containers, there will essentially be a lot of containers each trying to update their own heartbeat, even though only one really owns the lock, and the rest of the updates are going to be a waste and create extra load on the database.

It may still be ok without this, but if it's easy to check that could save a lot of unnecessary db update attempts.